### PR TITLE
fix(tests): derive releasepolicy script deadlines from test timeout

### LINKED ISSUE #1974: tests/releasepolicy: 10s per-script deadline flakes under full-suite load on dev machines
## Summary

`tests/releasepolicy`'s shell-script tests enforce a hardcoded 10-second wall-clock deadline per script invocation (`runScript`, `release_scripts_test.go:1425`). Under full-suite `make test` parallelism (`-race`, `-covermode=atomic`, ~96 packages concurrently), individual script runs starve past that margin on developer machines and fail with `script exceeded test deadline: context deadline exceeded` — a false alarm that intermittently breaks the documented pre-push `make qualify` gate.

## Evidence (2026-07-30, Apple Silicon dev machine)

- Failed 3 of 7 full local `make qualify` runs in one day — with a **different subset of tests each time** (e.g. `TestReleasePromotionPrereleaseWritesOnlyVersionAliases`, `TestReleasePreflightPaginatesAndPeelsAnnotatedTags` in one run; `TestReleaseTargetRerunPolicy`, `TestReleaseStablePromotionConvergesAndRerunsWithoutMutation` in another). Every failure is the deadline sentinel, never a functional assertion.
- The package **passes in isolation** on the same machine with the gate's exact flags and env (`AICR_CRITERIA_STRICT=1 go test -short -count=1 -race`), ~153s total across all 31 script invocations.
- Clean `origin/main` in-suite on the same machine: passed, but at 190s package time — visibly near the margin even without any in-flight change.
- The triggering PR branch (#1967) touches nothing the package executes; its only contribution is more concurrent test load elsewhere in the suite.
- Never observed on CI runners (checks green throughout).

## Root cause

A fixed wall-clock deadline measures the machine and its concurrent load, not the script's correctness. The deadline exists to catch a *hung* script (infinite loop, stuck prompt) — and hangs are separated from slow-but-working runs by orders of magnitude, so a tight constant buys no detection value while making the gate environment-sensitive.

## Proposed fix

Derive the per-script budget from the test binary's own deadline instead of a constant, with a hang-detection-sized cap:

```go
budget := 90 * time.Second
if d, ok := t.Deadline(); ok {
    if remaining := time.Until(d) - 30*time.Second; remaining < budget {
        budget = remaining
    }
}
ctx, cancel := context.WithTimeout(context.Background(), budget)
```

This inherits whatever `-timeout` the environment (gate, CI, dev machine) was provisioned with, still kills genuine hangs with margin left to report cleanly, and needs no per-machine tuning. The simple alternative — raising the constant to 60–120s — removes the flake equally well but leaves a constant to re-tune.

Priority: P3 (low) — dev-machine gate noise only; not observed on CI.

### PR DESCRIPTION
**Pull request by yuanchen8911**
## Summary

Replace the hardcoded 10-second per-script deadlines in `tests/releasepolicy` with a budget derived from the test binary's own deadline (`t.Deadline()`), capped at a hang-detection-sized 90 seconds.

## Motivation / Context

The fixed 10s wall-clock deadline measures the machine and its concurrent load, not the script's correctness. Under full-suite `make test` parallelism (`-race`, `-covermode=atomic`), individual script runs starve past that margin on developer machines and fail with `script exceeded test deadline` — intermittently breaking the documented pre-push `make qualify` gate. The deadline exists to catch a *hung* script, and hangs sit orders of magnitude above slow-but-working runs, so a tight constant buys no detection value.

Fixes: #1974
Related: #2056

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [x] Other: `tests/releasepolicy` (release script shell tests)

## Implementation Notes

- `scriptBudget(t)` grants `max(remaining − 30s margin, remaining/2)`, capped at a 90s hang-detection ceiling (`scriptHangCap`). The margin (`deadlineReportMargin`) is reserved so a killed script still fails via `t.Fatalf` with its captured output instead of the test runner's deadline panic — but never more than half the remaining time is taken as margin, so a short targeted `-timeout` (e.g. 25s) still yields a usable budget instead of collapsing to a false failure.
- The budget never exceeds the real remaining time, and once the deadline has passed it goes non-positive: the context expires immediately, failing fast with the script-timeout message instead of starting doomed work.
- The calculation is a pure helper, `scriptBudgetFor(remaining)`, with `scriptBudget(t)` keeping only the `t.Deadline()` handling — pinned by a table-driven boundary test (`TestScriptBudgetFor`: hang cap, cap boundary, full margin, half-margin cap, short `-timeout`, nearly exhausted, expired).
- All three subprocess helpers in the package share the derived budget: `runScript` (~31 invocations), `shellReleaseAssetNames`, and `runOpenVEXGuard`. No fixed-deadline subprocess sites remain in the package.
- The budget inherits whatever `-timeout` the environment (gate, CI, dev machine) was provisioned with — no per-machine constant to re-tune. Timeouts use named constants per the #2056 convention.

## Testing

```bash
AICR_CRITERIA_STRICT=1 go test -short -count=1 -race ./tests/releasepolicy/   # ok, ~170s
golangci-lint run -c .golangci.yaml ./tests/releasepolicy/...                 # 0 issues
# short targeted -timeout must not collapse the budget (review-round repro):
go test -count=1 -run '^TestReleaseStablePromotionConvergesAndRerunsWithoutMutation$' \
  -timeout=25s ./tests/releasepolicy                                          # ok, ~12s
make qualify
```

Coverage gate: test-only change in a test-only package (`_test.go`); no exported production code affected, no coverage delta.

## Risk Assessment

- [x] **Low** — Isolated change, well-tested, easy to revert

**Rollout notes:** N/A — test-infrastructure only; no user-facing behavior.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [ ] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`) — [GPG signing info](https://docs.github.com/en/authentication/managing-commit-signature-verification)

### CODE REVIEW

File: tests/releasepolicy/release_scripts_test.go [hunk 1]
```diff
@@ -1677,9 +1678,46 @@ type scriptResult struct {
...
+	remaining := time.Until(deadline)
+	budget := remaining - deadlineReportMargin
+	if half := remaining / 2; budget < half {
+		budget = half
+	}
+	if budget > scriptHangCap {
+		budget = scriptHangCap
+	}
+	return budget
+}
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply tests/releasepolicy/release_scripts_test.go:1722]: Addressed in 6038fce27: the calculation is extracted into a pure helper `scriptBudgetFor(remaining time.Duration)` (with `scriptBudget(t)` keeping only the `t.Deadline()` handling), pinned by the table-driven `TestScriptBudgetFor` covering ample time → 90s hang cap, the exact cap boundary, full 30s margin reserved, margin capped at half the remaining time, a short targeted `-timeout` (25s → 12.5s), a nearly exhausted deadline (1s → 500ms), and an expired deadline yielding a non-positive budget for immediate fail-fast.

File: tests/releasepolicy/release_scripts_test.go [hunk 2]
```diff
@@ -1677,9 +1678,74 @@ type scriptResult struct {
...
+func TestScriptBudgetFor(t *testing.T) {
+	tests := []struct {
+		name      string
+		remaining time.Duration
+		expected  time.Duration
+	}{
+		{"ample time hits hang cap", 10 * time.Minute, scriptHangCap},
+		{"cap boundary", scriptHangCap + deadlineReportMargin, scriptHangCap},
+		{"full margin reserved", 100 * time.Second, 70 * time.Second},
+		{"margin capped at half", 2 * deadlineReportMargin, deadlineReportMargin},
```
Comment (njhensley) [inline tests/releasepolicy/release_scripts_test.go]: **🔵 Nitpick — "margin capped at half" case doesn't exercise the half-bump branch**

At remaining=60s (2×deadlineReportMargin), budget = 60-30 = 30s and half = 30s are equal, so `budget < half` is false — the half-bump branch does not fire; the 30s result comes from the full-margin path. The case name implies it validates the half-cap regime, but that regime is actually exercised only by the 25s and 1s rows. Not a coverage gap (the branch is covered), just a naming/clarity mismatch.

**Blast radius:** Test readability only.

**Fix:** Optional: add an interior half-regime row, e.g. remaining=50s → budget=20 < half=25 → bumped to 25s, so the half-bump is the load-bearing assertion; or rename this case to 'full-margin/half tie at 2×margin'.

File: tests/releasepolicy/release_scripts_test.go [hunk 3]
```diff
@@ -1677,9 +1678,74 @@ type scriptResult struct {
...
+				t.Errorf("scriptBudgetFor(%v) = %v, want %v", tt.remaining, got, tt.expected)
+			}
+		})
+	}
+}
+
 func runScript(t *testing.T, environment []string, relative string, args ...string) scriptResult {
 	t.Helper()
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), scriptBudget(t))
```
Comment (njhensley) [inline tests/releasepolicy/release_scripts_test.go]: **🔵 Nitpick — script-kill message omits the derived budget**

When a script is killed, the failure reports ctx.Err() ('context deadline exceeded') but not the budget in effect. At the default 10m -timeout the budget is always 90s, so a kill means a real hang and the message is actionable. If the package is ever run under a short -timeout, scriptBudgetFor shrinks the budget below 90s and the same message can no longer tell an operator whether the script hung (~90s) or was starved by a tight budget.

**Blast radius:** Cosmetic under current config; a diagnosis aid only in the (currently non-existent) short-timeout scenario.

**Fix:** Optional: compute `budget := scriptBudget(t)` once, pass it to both WithTimeout and the message, e.g. t.Fatalf("script exceeded %s budget (from -timeout): %v\n%s", budget, ctx.Err(), output).

File: tests/releasepolicy/release_scripts_test.go [hunk 4]
```diff
@@ -1677,9 +1678,74 @@ type scriptResult struct {
...
+// deadlineReportMargin is reserved from the test binary's -timeout so a
+// killed script still fails via t.Fatalf with its captured output instead
+// of being torn down by the test runner's deadline panic.
+const deadlineReportMargin = 30 * time.Second
+
+// scriptBudget derives the per-script deadline from the test binary's own
+// deadline so the budget inherits whatever -timeout the environment (gate,
+// CI, dev machine) was provisioned with, capped at hang-detection scale.
+func scriptBudget(t *testing.T) time.Duration {
+	t.Helper()
```
Comment (njhensley) [inline tests/releasepolicy/release_scripts_test.go:1697]: **🔵 Nitpick — t.Helper() in scriptBudget is a no-op**

scriptBudget never calls t.Fatal/t.Error, so t.Helper() has no effect on failure-line attribution here. Harmless and conventional (matches other *testing.T helpers in the file).

**Blast radius:** None.

**Fix:** None required; leave as-is for consistency, or drop the t.Helper() call.

**Review by njhensley — DISMISSED** [review]
## Multi-persona review — recommendation: ✅ Approve with comments

**Method:** independent persona reviewers (correctness / test-design, CI-DX / operability, Go idioms) → adversarial re-derivation against the resolved code. The budget math was re-verified by hand across all seven boundary cases; `TestScriptBudgetFor` passes, `go vet` and `golangci-lint -c .golangci.yaml ./tests/releasepolicy/...` are clean.

**Tier legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

This is a clean, well-factored, test-only fix for #1974. The pure/impure split (`scriptBudgetFor` testable, `scriptBudget` isolating `t.Deadline()`) is idiomatic; naming and comments are accurate; the removed `time` import is genuinely unused; test-local constants are the right call (the package imports `pkg/defaults` nowhere and these describe test-runner deadline arithmetic, not a production default). Under `make test`/`make qualify` (`test_timeout: 10m`) every one of the ~31 `runScript` calls receives the full 90s `scriptHangCap` — 9× the old fixed 10s — so #1974's dev-machine starvation is no longer reachable, with no CI regression.

The core safety property holds: for every positive `remaining`, the derived budget is strictly `< remaining`, so the context timeout always fires before the test runner's own deadline panic, leaving margin for `t.Fatalf` to report captured output. Expired/negative deadlines fail fast correctly.

### 🟡 Minor — PR-body completeness claim is inaccurate; two same-class sites remain unconverted

The description states *"All three subprocess helpers in the package share the derived budget … No fixed-deadline subprocess sites remain in the package."* Two more subprocess helpers in the same package still use hardcoded `context.WithTimeout` deadlines on structurally identical `exec.CommandContext(ctx, "bash", …)` step-runners:

- [`release_reverify_test.go:1256`](https://github.com/NVIDIA/aicr/blob/6038fce27da1335ff00109b4218996b17fef2c6b/tests/releasepolicy/release_reverify_test.go#L1256) — `runReverifyCloseStep`, `30*time.Second`
- [`release_reverify_test.go:1486`](https://github.com/NVIDIA/aicr/blob/6038fce27da1335ff00109b4218996b17fef2c6b/tests/releasepolicy/release_reverify_test.go#L1486) — `runReverifyClassifier`, `60*time.Second`

These are the *same class* of flake #1974 describes. Their larger 30s/60s margins make recurrence materially less likely, and #1974 did not demonstrate a flake at these sites, so this is **non-blocking** — but the "no fixed-deadline sites remain" defense-in-depth the body claims isn't actually true. Either convert both to `scriptBudget(t)` for a uniform fix (one-line swap per site), or narrow the PR-body claim to the three script/guard helpers and note these two as out of scope. Not a merge blocker.

### Inline nitpicks (🔵)
Three optional polish items posted inline: a test-table case whose name over-promises (half-bump branch is a tie, not exercised, at 60s — the branch is covered elsewhere), the kill message omitting the derived budget (cosmetic at the 10m timeout), and a no-op `t.Helper()`. None require changes.

### Confirmed non-issues (examined and cleared)
- Budget exceeding real remaining time → impossible for positive input (proven per-branch).
- Half-margin reserving too little for `t.Fatalf` → microsecond operation; 500ms is ample.
- `ctx.Err()` post-`CombinedOutput` timeout detection → canonical Go idiom, **pre-existing** (not introduced here).
- Hand-computed expected values in the test table → intentional independent oracle.
- Cosign shim comment rewrite → accurate; correctly generalizes the old fixed-10s claim.

### Notes
- The author (`yuanchen8911`) shows as `COMMENTED` on this PR — that is a self-comment, not an independent reviewer; no other human review to duplicate.
- CodeRabbit's earlier `<1s`-fallback concern is already resolved by the half-margin clamp (marked addressed in `a2fce81`).

| 🔴 Blocker | 🟠 Major | 🟡 Minor | 🔵 Nitpick | Recommendation |
|-----------|---------|---------|-----------|----------------|
| 0 | 0 | 1 | 3 | ✅ Approve with comments |

File: tests/releasepolicy/release_scripts_test.go (hunk 2 above)
Comment (yuanchen8911) [reply tests/releasepolicy/release_scripts_test.go]: Fixed in 999d68b5d: renamed the 60s case to "full-margin/half tie at 2x margin" and added an interior half-regime row (50s -> 25s, where budget = 20s < half = 25s bumps), so the half-bump branch is now the load-bearing assertion of its own named case.

File: tests/releasepolicy/release_scripts_test.go (hunk 3 above)
Comment (yuanchen8911) [reply tests/releasepolicy/release_scripts_test.go]: Fixed in 999d68b5d: the budget is computed once and included in the kill message — "script exceeded 90s budget (derived from -timeout): ..." — in runScript, and the same treatment applied to shellReleaseAssetNames and runOpenVEXGuard so all three subprocess kill sites report the budget in effect.

File: tests/releasepolicy/release_scripts_test.go (hunk 4 above)
Comment (yuanchen8911) [reply tests/releasepolicy/release_scripts_test.go:1697]: Agreed it's a no-op — keeping t.Helper() as-is for consistency with the other *testing.T helpers in the file, per the "leave as-is" option.

**Review by njhensley — APPROVED** [review]
